File: 프로그래머스/lv2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;

class Solution {
    int[] dxs = {1, -1, 0, 0}, dys = {0, 0, 1, -1};
    int n, m;
    
    public boolean inRange(int x, int y){
        return 0 <= x && x < m && 0 <= y && y < n;
    }
    
    public int solution(int[][] maps) {
        n = maps.length;
        m = maps[0].length;
        
        int[][] visited = new int[n][m];
        for (int i = 0; i < n; i++){
            for (int j = 0; j < m; j++)
                visited[i][j] = n * m + 1;
        }
        
        Deque<Pos> dq = new LinkedList<>();
        dq.addFirst(new Pos(0, 0, 1));
        
        while (!dq.isEmpty()){
            Pos cur = dq.poll();
            for (int i = 0; i < 4; i++){
                int nx = cur.x + dxs[i];
                int ny = cur.y + dys[i];
                int nz = cur.z + 1;
                
                if (inRange(nx, ny) && maps[ny][nx] == 1 && visited[ny][nx] > nz){
                    dq.add(new Pos(nx, ny, nz));
                    visited[ny][nx] = nz;
                }
            }
        }
        
        
        
        return visited[n - 1][m - 1] == n * m + 1 ? -1 : visited[n - 1][m - 1];
    }
}

class Pos{
    public int x;
    public int y;
    public int z;
    public Pos(int x, int y, int z){
        this.x = x;
        this.y = y;
        this.z = z;
    }
}

File: 프로그래머스/lv2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque
dxs, dys = (1, -1, 0, 0), (0, 0, 1, -1)


def solution(maps):
    dq = deque()
    n, m = len(maps), len(maps[0])
    distance = [[n * m + 1 for _ in range(m)] for __ in range(n)]
    distance[0][0] = 1
    dq.appendleft((0, 0, 1))
    while dq:
        x, y, dis = dq.pop()
        for dx, dy in zip(dxs, dys):
            nx, ny = x + dx, y + dy
            if 0 <= nx < m and 0 <= ny < n and maps[ny][nx] and distance[ny][nx] > dis + 1:
                distance[ny][nx] = dis + 1
                dq.appendleft((nx, ny, dis + 1))
    return distance[n - 1][m - 1] if distance[n - 1][m - 1] != n * m + 1 else -1